src/main/java/com/faforever/api/deployment/ExeUploaderController.java

```diff
@@ -0,0 +1,85 @@
+import com.faforever.api.error.ApiException;
+import com.faforever.api.error.Error;
+import com.faforever.api.error.ErrorCode;
+import com.faforever.api.featuredmods.FeaturedModFile;
+import com.faforever.api.featuredmods.FeaturedModService;
+import com.google.common.io.Files;
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.security.core.Authentication;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.LinkedList;
+import java.util.List;
+
+import static com.google.common.hash.Hashing.md5;
+import static com.google.common.io.Files.hash;
+import static org.springframework.http.MediaType.APPLICATION_JSON_UTF8_VALUE;
+
+@RestController
+@RequestMapping(path = "/exe")
+@Slf4j
+public class ExeUploaderController { // basic
+  private final FafApiProperties apiProperties;
+  private final FeaturedModService featuredModService;
+  private final ExeUploaderService exeUploaderService;
+
+  public ExeUploaderController(
+    FafApiProperties apiProperties,
+    FeaturedModService featuredModService,
+    ExeUploaderService exeUploaderService
+  ) {
+    this.apiProperties = apiProperties;
+    this.featuredModService = featuredModService;
+    this.exeUploaderService = exeUploaderService;
+  }
+
+  @ApiOperation("Upload an exe file")
+  @ApiResponses(value = {
+    @ApiResponse(code = 200, message = "Success"),
+    @ApiResponse(code = 401, message = "Unauthorized"),
+    @ApiResponse(code = 500, message = "Failure")})
+  @RequestMapping(path = "/upload", method = RequestMethod.POST, produces = APPLICATION_JSON_UTF8_VALUE)
+  public void upload(@RequestParam("file") MultipartFile file,
+                     @RequestParam("modName") String modName,
+                     Authentication authentication
+  ) throws IOException {
+    // TODO: auth
+    if (file == null) {
+      throw new ApiException(new Error(ErrorCode.UPLOAD_FILE_MISSING));
+    }
+    if (modName == null) {
+      throw new ApiException(new Error(ErrorCode.PARAMETER_MISSING, "modName"));
+    }
+    String extension = Files.getFileExtension(file.getOriginalFilename());
+    if (!apiProperties.getExe().getAllowedExtensions().contains(extension)) {
+      throw new ApiException(new Error(ErrorCode.UPLOAD_INVALID_FILE_EXTENSIONS, apiProperties.getExe().getAllowedExtensions()));
+    }
+
+    log.info("Starting deployment modName '{}', file '{}'", modName, file.getOriginalFilename());
+
+    updateDatabase(file, modName);
+  }
+
+  private void updateDatabase(MultipartFile file, String modName) throws IOException {
```

Comment:
The controller class should not perform anything. It should be only used as the proxy for the services defining the URLs, describing everything with Swagger annotations and checking security stuff. 

Also there should be overlay with method `updateDatabase` in class `LegacyFeaturedModDeploymentTask`. If necessary, shared code could be mode to the `FeaturedModService`.

Alternatively it belong to the `ExeUploaderService`

Reply:
I'll move stuff to `ExeUploaderService`